### DISCUSSION TITLE
Release of invalid GC handle. The handle is from a previous domain. The release operation is skipped.

### DISCUSSION THREAD

**Discussion by tunyawat** [RageAgainstThePixel/com.openai.unity#209] [Q&A] [answered]
First of all, I want to thank you for creating such a helpful Unity asset. It has significantly contributed to the completion of my project.

However, I've encountered an odd issue after using the asset. Occasionally, messages from the Assistant are sent twice from a single chat request. The content of the messages is different.  There are sometimes, a message is sent but receive no reply. 

When such issues arise, Unity has reported an error stating, "Release of invalid GC handle. The handle is from a previous domain. The release operation is skipped."

Here is the code I've implemented in my Unity project. I noticed that when the Assistant sends a message twice, only one is recorded in the List<message>.

```csharp
    public async void AskChatGPTAsync(string newText)
    {
        responseTMP.text = "";
        loadingAnimation.SetActive(true);      

        messages.Add(new Message(Role.User, newText));

        var chatRequest = new ChatRequest(messages, languageModel);
        var response = await api.ChatEndpoint.GetCompletionAsync(chatRequest);
        var choice = response.FirstChoice;

        loadingAnimation.SetActive(false);

        responseTMP.text = choice.Message;
        messages.Add(choice.Message);

        foreach (var message in messages)
        {
            Debug.Log($"{message.Role}: {message}");
        }
    }
```

**Comment by StephenHodgson** [comment, marked as answer]
Hi @tunyawat. I'll need better steps to reproduce this issue you're running into.

It sounds like maybe are reloading the domain or entering and exiting from play mode but not cancelling your previous sessions call.

You should also be passing a cancellation token source. You can create your own or use [`GameObject.destroyCancellationToken`](https://docs.unity3d.com/ScriptReference/MonoBehaviour-destroyCancellationToken.html)

```csharp
var response = await api.ChatEndpoint.GetCompletionAsync(chatRequest, destroyCancellationToken);
```

**Comment by tunyawat** [comment]
I will try and give the feedback later, thanks.

**Reply by tunyawat** [reply]
By implementing the code based on your suggestion, the issue has been resolved successfully. Thank you for your guidance!

Here are my code snippets (just in case others may find them useful).
```csharp
var response = await api.ChatEndpoint.GetCompletionAsync(chatRequest, destroyCancellationToken);
```
